OpenSim/Common/Test/testOperators.cpp

```diff
@@ -0,0 +1,421 @@
+        addComponent(&_comVelocityDelay);
+    }
+    void extendConnectToModel(Model& model) override {
+        Super::extendConnectToModel(model);
+        const auto& coord = model.getCoordinateSet()[0];
+        getInput("coord_value").connect(coord.getOutput("value"));
+        _delay.getInput("input").connect(coord.getOutput("value"));
+        _vectorDelay.getInput("input").connect(getOutput("coord_value_vector"));
+        _comVelocityDelay.getInput("input").connect(
+                model.getOutput("com_velocity"));
+    }
+
+    ScalarDelay _delay;
+    VectorDelay _vectorDelay;
+    Delay<SimTK::Vec3> _comVelocityDelay;
+};
+
+/// Contains dummy outputs with different requiresAt stages so that we can
+/// test that stages are checked appropriately when asking for the Delay output.
+class DelayStageTesting : public ModelComponent {
+OpenSim_DECLARE_CONCRETE_OBJECT(DelayStageTesting, ModelComponent);
+public:
+    DelayStageTesting() {
+        constructInfrastructure();
+    }
+    // This also tests the convenience constructor.
+    ScalarDelay posDelay{0.01};
+    ScalarDelay velDelay{0.02};
+    ScalarDelay dynDelay{0.03};
+
+private:
+    void constructOutputs() override {
+        constructOutput<double>("position",
+                std::bind([](const SimTK::State& s)->double
+                        { return s.getTime() + 1.0; },
+                        std::placeholders::_1),
+                SimTK::Stage::Position);
+        constructOutput<double>("velocity",
+                std::bind([](const SimTK::State& s)->double
+                        { return s.getTime() + 2.0; },
+                        std::placeholders::_1),
+                SimTK::Stage::Velocity);
+        /* TODO
+        There's a bug where outputs with a dependsOn stage of Dynamics or
+        above causes an exception to be thrown when initializing the
+        TimeStepper.
+        constructOutput<double>("dynamics",
+                std::bind([](const SimTK::State& s)->double
+                        { return s.getTime() + 3.0; },
+                        std::placeholders::_1),
+                SimTK::Stage::Dynamics);
+        */
+    }
+    void extendFinalizeFromProperties() override {
+        Super::extendFinalizeFromProperties();
+        addComponent(&posDelay);
+        addComponent(&velDelay);
+        //TODO see above addComponent(&dynDelay);
+    }
+    void extendConnectToModel(Model& model) override {
+        Super::extendConnectToModel(model);
+        posDelay.getInput("input").connect(getOutput("position"));
+        velDelay.getInput("input").connect(getOutput("velocity"));
+        //TODO see above dynDelay.getInput("input").connect(getOutput("dynamics"));
+    }
+};
+
+/// Allows some testing of Delay with different models.
+void testDelaySimulation(Model& model, double delayTime = 0.017,
+        bool testRegression = true) {
+
+    double finalTime = 1.35;
+    double tol = 1e-6; // integrator accuracy, test tolerance.
+
```

Comment:
What happens when the integration step size is larger than the delay? It would be good to force this case possibly by i) reducing the delay time and ii) loosening the tolerance.